### ISSUE TITLE
dirname of current Taskfile

### ISSUE THREAD

**Issue by fj** [go-task/task#215]
I often want to use the directory of where the Taskfile is located as an input to other commands (not the current working directory), but it seems like `sprig` really locks down access to the filesystem.

Is this possible to do with `Task` as a first-class idea?

For example, something like:

```
tasks:
  show-current-dir:
    desc: "show where Task is executing from"
    cmds:
      - echo {{ .TASKFILE_LOCATION }}
```

Note that this isn't the same as just using `$PWD` or `-d` because the working directory might be different; Task should probably provide this as a first-class feature if it doesn't already.

**Comment by andreynering** [comment]
Hi @fj,

You should actually be fine using `pwd` if you haven't specified another `dir:` or your haven't `cd`ed in the command.

Otherwise, yeah, I think it makes sense to a have variable with the Taskfile location.

**Comment by smyrman** [comment]
Actually, could be handy for when Taskfiles are imported as well.

**Comment by smyrman** [comment]
Example use-case:

root:
```yaml
version: "2.X"
include:
  sub: sub/Tasks.yaml
tasks:
  default:
    cmds:
    - sub:test
```
root/sub:
```yaml
version: "2.X"
tasks:
  test:
    dir: "{{.TASKFILE_DIR}}"
    cmds:
    - ./bin/executable
```
root/sub/bin/executable
```sh
#!/bin/sh
echo "Hello World"
```

**Comment by beevelop** [comment]
Just stumbled upon this exact issue. I guess a workaround for now is manually setting the `dir` to the respective folder. The negative side-effect: The `Taskfile` in the subfolder can no longer be used independently.

**Comment by max-sixty** [comment]
Is there a way of getting the current working directory when a taskfile in another dir is invoked with `-t`?

e.g. a variable that would return `/home/foo` here:
```sh
$ pwd

/home/foo

$ task -t /home/foo/bar/Taskfile-baz.yml

# assuming task is `echo {{ .PWD }}`

/home/foo/bar
```

**Comment by SEAPUNK** [comment]
I'm curious as to why this isn't the default behavior. If I had a setup like

```yml
# Taskfile.yml
version: '3'
includes:
  db: ./db

# db/Taskfile.yml
version: '3'
tasks:
  something: do_something_with ./file.sql
```

I would expect to be able to do both `task db:something` and `cd db; task something` and have them both work. Why doesn't pwd get set per-taskfile? Why do I have to work around this manually?

**Comment by andreynering** [comment]
@SEAPUNK It seems to be that you just forgot to specify the dir of the Taskfile:

```yaml
includes:
  db:
    taskfile: ./db
    dir: ./db
```

To the others: I know that this issue has been opened for a long time (I have been having less than enough time to maintain this), but I'll try to give this some priority.

**Comment by SEAPUNK** [comment]
@andreynering no, that's what I mean - why should I have to do that? It doesn't feel very intuitive to me (especially considering the quote from the usage page: "By default, tasks will be executed in the directory where the Taskfile is located."), and I end up having to manually set the dir for every taskfile that I import.

**Comment by fj** [comment]
Thanks for the update, Andrey.

As the person who opened the issue, I do want to acknowledge that as this
is a free and open-source project, I don't have any expectation that you
have a timetable for fulfilling requests or fixing issues, and that any
time you devote at all is appreciated.

best,
~ jf

**Comment by andreynering** [comment]
@fj Thanks for the words! 😉

---

@SEAPUNK There's some historical reasons for why it works as is today. Originally, including other taskfiles was a simpler feature, and it was impossible to make it directory aware. When it was added (the ability to set a "dir" for a taskfile import), it was made in a backward compatible way.

Changing it now would break many taskfiles in the wild, because I do know many store common taskfiles in a given directory (I've seen a few `.taskfiles/` and similar in the wild) without waiting for these tasks to really run on this folder, but on root.

**Comment by SEAPUNK** [comment]
@andreynering gotcha, thanks for the explanation. I thought this wasn't something that showed up in the docs, but it looks like I skipped over it.

**Comment by rwxd** [comment]
> Is there a way of getting the current working directory when a taskfile in another dir is invoked with `-t`?
> 
> e.g. a variable that would return `/home/foo` here:
> 
> ```shell
> $ pwd
> 
> /home/foo
> 
> $ task -t /home/foo/bar/Taskfile-baz.yml
> 
> # assuming task is `echo {{ .PWD }}`
> 
> /home/foo/bar
> 
> # how could we get /home/foo ?
> ```

Hi @max-sixty, have you found a solution to the problem?

**Comment by max-sixty** [comment]
> Hi @max-sixty, have you found a solution to the problem?

I found another way of doing what I needed, sorry not to be helpful.

**Comment by tylermmorton** [comment]
Hi, is it OK to close this issue with the provided workarounds?

**Comment by neerfri** [comment]
@tylermmorton I think the need for a variable that allows getting the directory of the Taskfile is still open and relevant despite the workarounds

**Comment by ssbarnea** [comment]
Anything involving PWD would not count as safe, IMHO.

@andreynering Sorry to nag you about this but is there any chance to close this before this bug reaches the 2nd year anniversary?  At least if there is a valid workaround we should document it, especially as the need to have such a variable is quite common.

I personally do not care about the effective taskfile location as at this moment I do not use inclusion but I do need a "project_dir" variable, one that would not change even if I call task from a subdirectory.

**Comment by ghostsquad** [comment]
I agree, task should be able to report directory in which the Taskfile it's operating on is located, though this brings into question, should it always be the parent Taskfile directory, and never the directory of the Taskfiles that are in `include`?

**Comment by kponichtera** [comment]
There are legitimate use cases where having access to the Taskfile's working directory would be useful - for example mounting the working directory to Docker container. When specifying volume, Docker requires an absolute path of the directory to be specified. The current workaround for this is to do as follows:

```yaml
tasks:
  run-in-docker:
    vars:
      PWD:
        sh: pwd
    cmds:
      - >
        docker run --rm
        --workdir /project
        --volume {{.PWD}}:/project
        busybox:latest
        ping google.com
```

However, the problem arises when there are two tasks like that, executed in parallel (by specifying them in their parent task `deps`) - in this case the `PWD` variable of one task replaces the value of the other one. Technically it is possible to just rename the variables, but that introduces a lot of noise for something as simple as fetching current directory. Being able to access current working directory of the Taskfile with something like `{{.WORKING_DIR}}` would solve that.

**Comment by andreynering** [comment]
@ghostsquad I believe we could have two variables: `ROOT_TASKFILE`, `CURRENT_TASKFILE` or something similar.

Sorry for taking so long to work on this, we just have too many opened issues to triage.

This seems to be easy and requested enough that will just put at the top of my priority list.

But if in the meantime, if someone else would like to contribute with this, just let me know.

**Comment by ssbarnea** [comment]
I was thinking about `.PROJECT_DIR` and `.TASKFILE_DIR`, with the mention that the last one implies "current taskfile dir" but I am sure I could live with any variable names, still try to avoid confusing user with a variable name ending in file (TASKFILE) if that is a directory ;)

**Comment by andreynering** [comment]
Ah, of course, I missed the `_DIR` prefix, we certainly need it.

My point was about having two variables.

👍 for `PROJECT_DIR` (alternatively `ROOT_DIR`) and `TASKFILE_DIR`

**Comment by andreynering** [comment]
This was just implemented as part of #857. I plan to do a release still today.
